### ISSUE TITLE
[java] ConfusingTernary should treat != null as positive condition

### ISSUE THREAD

**Issue by sebbASF** [pmd/pmd#278]
**Rule Set:** [ConfusingTernary](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#confusingternary)

**Description:** 

ConfusingTernary should treat negative comparisons with null as a positive condtion.

The condition "!= null" is often used to mean "the item is present" rather than "the item is not absent".

Rewriting using "== null" would make the code much harder to follow.

**Code Sample demonstrating the issue:**

```
if (parser.getArgumentById(VERSION_OPT) != null) {
...
} else if (parser.getArgumentById(HELP_OPT) != null) {
...
} else if (parser.getArgumentById(OPTIONS_OPT) != null) {
...
} else if (parser.getArgumentById(SERVER_OPT) != null) {
...
```

**Running PMD through:** *Ant*

**Comment by jsotuyod** [comment]
@sebbASF thanks for the report!

I think there are a bunch of things to think about on this most interesting report.

Firstly, regardless of all else, the rule is certainly just introducing noise. The rule attempts to avoid "inverted" conditionals such as *"if NOT A then B else C"*, but on this scenario, all checks are for different things altogether. The rule should not apply to this block.

Right now, there is a `ignoreElseIf` property for the rule, to have it ignore all `else if`. This would remove this warning, but will certainly be broader than expected.

Then it's the point of considering "!=" a positive. That would allow to write code such as:

```java
if (x != null) {
  return x.foo();
} else {
  return 0;
}
```

which is the very kind of code this rule attempts to prevent.


What I guess should happen is:
 - the rule should be smart enough to check if `else if` is concerning the same variable, and only apply if so
 - the `ignoreElseIf` is probably redundant in such scenario, analyze removing it

**Comment by sebbASF** [comment]
My example happened to use the same variable, but there are other cases such as a hierarchy of defaults where the conditions need to check for the first non-null variable.

My point is that null is special here, because the natural check for non-null uses inequality.

**Comment by SUPERCILEX** [comment]
I'm having the same issue:
```java
if (deepLink.getQueryParameter(TEAM_QUERY_KEY) != null) { // PMD shouldn't report this as an error
    launchTeam(getTeam(deepLink).get(0));
} else if (deepLink.toString().equals(ADD_SCOUT_INTENT)) {
    NewTeamDialog.show(mActivity.getSupportFragmentManager());
}
```

**Comment by jsotuyod** [comment]
@SUPERCILEX yeap, similar case, you are analyzing different values on the ifs (all related to `deepLink`, but different all the same.

**Comment by ThrawnCA** [comment]
I think that *any* usage of `else if` should be enough to disable the rule. As soon as it's more complex than the binary case, there may be good reasons for writing it with the negative first.

**Comment by Oowekyala** [comment]
@jsotuyod 
> Then it's the point of considering "!=" a positive. 

I think the point is only to consider `!= null` as a positive. The rule will still catch confusing conditionals like
```java
if (x != 2) {
  //...
} else {
  //...
}
```
But null checks will be ignored because it's (arguably) as easy to read `if (x != null)` as to read `if (x == null)`.

**Comment by sebbASF** [comment]
Yes; see the title and initial comment of this issue.

**Comment by alsingh287** [comment]
Any updates on this ?

**Comment by sabberworm** [comment]
I would also agree that a null check is a special case that needs to be treated differently.

if(something is null) {
  do the fallback
} else {
  do the thing with something
}

is much more confusing to read than

if(something isn’t null) {
  do the thing with something
} else {
  do the fallback
}

**Comment by JoyChopra1298** [comment]
Is this being considered? Any update?
